### DISCUSSION TITLE
Qustion about python scripting

### DISCUSSION THREAD

**Discussion by aminabedi68** [fontforge/fontforge#5334] [Q&A]
Hi
Is there a boolean function to check if either a point is or isn't inside the final black shape of a glyph?

**Comment by rudransh61** [comment]
```python
from shapely.geometry import Point, Polygon
glyph_shape = Polygon([(x1, y1), (x2, y2), ...])
point_to_check = Point(x, y)
is_inside = glyph_shape.contains(point_to_check)
print(is_inside)
```

**Comment by rudransh61** [comment]
You can try this function..

**Reply by aminabedi68** [reply]
thank you @rudransh61 i wrote a test, it's near but not working as expected for me. there is one issue:
i don't know polygon model could handle the bezier curve or not, so it seems the accuracy gets low in low-point curvy shapes.
i wrote a test with this file:
[test.zip](https://github.com/fontforge/fontforge/files/13835255/test.zip)
contains this glyph:
![image](https://github.com/fontforge/fontforge/assets/25493297/8204b979-596c-4ac2-b4f3-ba1d4808ea0d)
this is a script to generate points of the shape(all on-curve and off-curve points):
```python
import os
import fontforge

font = fontforge.open('test.sfd')
glyph = font['zero']
for contour in font[glyph.glyphname].layers[glyph.activeLayer]:
    points = list(contour)
    polygon_bezier = [(points[i].x,points[i].y) for i in range(len(points))]
    print(polygon_bezier)
```
and the check:
```python
from shapely.geometry import Point, Polygon

polygon_bezier = [(615.0, -11.0), (305.0, -11.0), (117.0, 258.0), (117.0, 703.0), (117.0, 1148.0), (304.0, 1416.0), (615.0, 1416.0), (925.0, 1416.0), (1111.0, 1149.0), (1111.0, 703.0), (1111.0, 257.0), (925.0, -11.0),
(908.0, 703.0), (908.0, 1049.0), (807.0, 1230.0), (615.0, 1230.0), (422.0, 1230.0), (320.0, 1048.0), (320.0, 703.0), (320.0, 358.0), (422.0, 175.0), (615.0, 175.0), (806.0, 175.0), (908.0, 358.0)]
glyph_shape = Polygon(polygon_bezier)
for point in [(0,0),(320,220),(615,703),(713,1245),(765,1210)]:
    print(point,glyph_shape.contains(Point(point)))
```

```batch
(0, 0) False
(320, 220) True
(615, 703) False
(713, 1245) True
(765, 1210) False
```
the last point is on the shape but ignored. there could be some mistakes but i think the reason for that is the polygon model with straight lines.

**Comment by skef** [comment]
I don't think FontForge does this by itself. You probably want https://fonttools.readthedocs.io/en/latest/pens/pointInsidePen.html from fontTools, which you should be able to get working with FontForge's pen interface. (At worst you might need to write a little adapter layer.)

**Reply by iorsh** [reply]
https://fontforge.org/docs/scripting/python/fontforge.html#fontforge.contour.draw draws a contour into a pen. FontTools pens appear compatible with RoboFab.

**Reply by aminabedi68** [reply]
thank you @skef this could be helpful.
I don't know the technical limitations, but I'm curious: why the same function doing plot the black area in gui preview or fill couldn't help to check the point?

**Reply by skef** [reply]
Yeah, you'd want to raster a big monochrome image in order to get decent accuracy and avoid fretting about grayscale anti-aliased values. But the accuracy would _only_ be "decent". It would be a bit of work to code up and then people would complain about wrong answers on the margin. Better to do it more analytically.

This isn't _that_ hard a problem to solve analytically, it's just something that doesn't happen to be in our source base.

**Reply by aminabedi68** [reply]
i wrote a script and it seems working fine. the algorithm is to draw a triangle with given to check point, and two point out of glyph bounding box then overlap with the original glyph outline(with unlinked references), and counting the difference of before and after list of full on-curve points :
```python
import os
import sys
import fontforge

font = fontforge.open(sys.argv[1])
glyph = font[sys.argv[2]]
glyph_bbox = glyph.boundingBox()
for ref in glyph.references:
	glyph.unlinkRef()
font.layers.add('temp_layer',False)
glyph.setLayer(glyph.layers['Fore'], 'temp_layer')
full_on_curve_points_list_before = []
full_on_curve_points_list_after = []
for contour in font[glyph.glyphname].layers[glyph.activeLayer]:
	points = list(contour)
	on_curve_points_list = [(points[i].x,points[i].y) for i in range(len(points)) if points[i].on_curve]
	for i in range(len(on_curve_points_list)):
		full_on_curve_points_list_before.append(on_curve_points_list[i])
point = (int(sys.argv[3]),int(sys.argv[4]))
pen = glyph.glyphPen()
pen.moveTo(point)
pen.lineTo((glyph_bbox[0]-10, glyph_bbox[1]-1))
pen.lineTo((glyph_bbox[0]-1, glyph_bbox[1]-10))
pen.lineTo(point)
pen.closePath()
glyph.setLayer(glyph.layers['temp_layer'] + glyph.layers['Fore'], 'Fore')
glyph.removeOverlap()
for contour in font[glyph.glyphname].layers[glyph.activeLayer]:
	points = list(contour)
	on_curve_points_list = [(points[i].x,points[i].y) for i in range(len(points)) if points[i].on_curve]
	for i in range(len(on_curve_points_list)):
		full_on_curve_points_list_after.append(on_curve_points_list[i])

if ((len(full_on_curve_points_list_after)-len(full_on_curve_points_list_before)-3)/4)%2:
	print(point,'True')
else:
	print(point,'False')
# font.save('test2.sfd') #to check what happened
```
testing:
```batch
REM fontforge -script test.py font glyph poin_x point_y
fontforge -script test.py test.sfd zero 765 1210
```
result:
```
(765, 1210) True
```
@skef i know this is stupid and you say no probably, just need to know: is it worth to add equivalent algorithm in fontforge api as a boolean function?(i didn't check in all conditions but it seems correct)

**Reply by skef** [reply]
@aminabedi68 what if the point is in the hole of an "O"?

**Reply by aminabedi68** [reply]
yes, this example(zero) has a hole. i wrote a test for guide layer(for list of test points inside glyph bounding box if the result is true then draw a tiny dimond in guide layer with center of that point):
```python
import os
import sys
import fontforge

font = fontforge.open(sys.argv[1])
glyph = font[sys.argv[2]]
glyph_bbox = glyph.boundingBox()
glyph_bbox_width = int(glyph_bbox[2] - glyph_bbox[0])
glyph_bbox_height = int(glyph_bbox[3] - glyph_bbox[1])
LSB = int(glyph.left_side_bearing)
resolution = int(sys.argv[3])
guides = font.guide
for ref in glyph.references:
    glyph.unlinkRef()
font.layers.add('temp_layer',False)
glyph.setLayer(glyph.layers['Fore'], 'temp_layer')
full_on_curve_points_list_before = []
for contour in font[glyph.glyphname].layers[glyph.activeLayer]:
    points = list(contour)
    on_curve_points_list = [(points[i].x,points[i].y) for i in range(len(points)) if points[i].on_curve]
    for i in range(len(on_curve_points_list)):
        full_on_curve_points_list_before.append(on_curve_points_list[i])
points = [(x,y) for x in range(LSB,glyph_bbox_width+LSB,resolution) for y in range(0,glyph_bbox_height,resolution)]
for point in points:
    full_on_curve_points_list_after = []
    pen = glyph.glyphPen()
    pen.moveTo(point)
    pen.lineTo((glyph_bbox[0]-10, glyph_bbox[1]-1))
    pen.lineTo((glyph_bbox[0]-1, glyph_bbox[1]-10))
    pen.lineTo(point)
    pen.closePath()
    glyph.setLayer(glyph.layers['temp_layer'] + glyph.layers['Fore'], 'Fore')
    glyph.removeOverlap()
    for contour in font[glyph.glyphname].layers[glyph.activeLayer]:
        points = list(contour)
        on_curve_points_list = [(points[i].x,points[i].y) for i in range(len(points)) if points[i].on_curve]
        for i in range(len(on_curve_points_list)):
            full_on_curve_points_list_after.append(on_curve_points_list[i])
    glyph.setLayer(glyph.layers['temp_layer'], 'Fore')
    if ((len(full_on_curve_points_list_after)-len(full_on_curve_points_list_before)-3)/4)%2:
        guide = fontforge.contour()
        guide.moveTo (point[0]-1, point[1])
        guide.lineTo (point[0], point[1]+1)
        guide.lineTo (point[0]+1, point[1])
        guide.lineTo (point[0], point[1]-1)
        guide.lineTo (point[0]-1, point[1])
        guide.closed = True
        guides += guide
        font.guide = guides
font.save('test2.sfd')
```

```batch
REM fontforge test.py glyphname resolution
fontforge -script test.py test.sfd zero 50
```
this is output:
![fontforge_KxDFDAc6Mx](https://github.com/fontforge/fontforge/assets/25493297/adb606e7-0f38-4dcb-8c26-06df5403baef)

with resolution = 10 :
![fontforge_OGZOLY73ay](https://github.com/fontforge/fontforge/assets/25493297/6e114fff-1acc-4ebc-9385-a1039fc723b5)

there is some inaccuracy in some points. i think the algorithm is correct but we need to check different points out of bounding box for triangle(inaccuracy almost happens in area that triangle side touching the inner contour)

what you think?

**Reply by skef** [reply]
Well, in terms of the correctness of the algorithm you would want to try things out with many more test cases, such as complex chinese characters with multiple holes.

As for the "include it in FontForge" question, you are probably aware that I see the goal of rewriting things in C for FontForge as being counter-productive and mostly exactly what FontForge doesn't need. This is the kind of thing that would make a great plugin though, if only we could get some momentum behind the plugin system. The functionality is there.

**Reply by iorsh** [reply]
Yes, the idea of plugins sound good. For now, maybe some Python scripts can be just put in a "Contrib" section

**Reply by skef** [reply]
No, no, the plugin infrastructure is _all done_. You can pip install a module that will then be "discovered" by FontForge, there's a UI for turning them on and off. The menu system was cleaned up and enhanced to work better (still under the "Tools" heading. There's a multi-question/answer dialog option. All this was added two versions ago, we just need people to start using it.

**Reply by skef** [reply]
https://fontforge.org/docs/techref/plugins.html
https://fontforge.org/docs/techref/pyextend.html
https://fontforge.org/docs/techref/askmulti.html

**Reply by skef** [reply]
Incidentally, @iorsh, the plugin stuff is also a lesson on what you should be worrying about.

I spent a good deal of time thinking about what would be the most effective route to making FontForge extendable without adding to the huge pile of problematic C code and then lots and lots and lots of time implementing all of this stuff. This, combined with its comprehensive python support, probably makes FontForge one of the more easily extendable UI applications out there. 

And basically no one knows, let alone cares. Open source work can be a black hole.

**Reply by iorsh** [reply]
Looks like it's very well hidden indeed. The link to pypi.org is broken, and even after fixing it I got ~2000 results, with only the first one looking relevant.

I would suggest to:
1. Add some preinstalled module, just to make the "Tools" menu look less sad and empty. This would clearly communicate users what "Tools" are for, without forcing them to go to the documentation.
2. Install existing modules automatically along with FF, at least until there are some 20-30 modules available (I'm too optimistic here probably). GIMP does that, I think.
3. Adding UI plugin manager when flat installation of everything becomes less feasible.

**Reply by skef** [reply]
The URL isn't being processed by the doc translator properly but the string "fontforge_plugin" is supposed to be in quotes. It's just that there aren't any so there's nothing in the pip database advertising that string. 

I was going to write a bunch based partly on issues that already have scripts as part of their answers, but I got a job and ran out of steam. 

In terms of automatic installation we talked some about that. It depends on what it does and who writes it. You want to avoid the "contrib directory problem", where person A write a script and hands it off to project B's contrib directory, which in practice means they can't easily edit it and the script rots.
